File: O-OpenClosed/Good_AreaCalculator.cs
// ✅ GOOD — Follows Open/Closed Principle
// The AreaCalculator never changes.
// To add a new shape, just CREATE a new class — no editing old code.

public abstract class Shape
{
    public abstract double Area();
}

public class Rectangle : Shape
{
    public double Width { get; set; }
    public double Height { get; set; }

    public override double Area() => Width * Height;
}

public class Circle : Shape
{
    public double Radius { get; set; }

    public override double Area() => Math.PI * Radius * Radius;
}

// ✅ New shape added WITHOUT touching any existing code
public class Triangle : Shape
{
    public double Base { get; set; }
    public double Height { get; set; }

    public override double Area() => 0.5 * Base * Height;
}

public class AreaCalculator
{
    // This method NEVER changes, no matter how many shapes you add
    public double Calculate(Shape shape) => shape.Area();
}

// Usage:
// var calc = new AreaCalculator();
// Console.WriteLine(calc.Calculate(new Rectangle { Width = 5, Height = 3 })); // 15
// Console.WriteLine(calc.Calculate(new Circle { Radius = 4 }));               // 50.26...
// Console.WriteLine(calc.Calculate(new Triangle { Base = 6, Height = 4 }));   // 12

File: O-OpenClosed/Good_AreaCalculator.cpp
// ✅ GOOD — Follows Open/Closed Principle
// AreaCalculator never changes.
// To add a new shape, just CREATE a new class — no editing old code.

#include <iostream>
#include <cmath>
#include <memory>
#include <vector>

// Abstract base
class Shape {
public:
    virtual double Area() const = 0;
    virtual ~Shape() = default;
};

class Rectangle : public Shape {
public:
    double Width, Height;
    Rectangle(double w, double h) : Width(w), Height(h) {}
    double Area() const override { return Width * Height; }
};

class Circle : public Shape {
public:
    double Radius;
    Circle(double r) : Radius(r) {}
    double Area() const override { return M_PI * Radius * Radius; }
};

// ✅ New shape added WITHOUT touching any existing code
class Triangle : public Shape {
public:
    double Base, Height;
    Triangle(double b, double h) : Base(b), Height(h) {}
    double Area() const override { return 0.5 * Base * Height; }
};

class AreaCalculator {
public:
    // This method NEVER changes, no matter how many shapes you add
    double Calculate(const Shape& shape) const { return shape.Area(); }
};

// Usage:
// AreaCalculator calc;
// std::cout << calc.Calculate(Rectangle{5, 3}) << "\n";  // 15
// std::cout << calc.Calculate(Circle{4})       << "\n";  // 50.26...
// std::cout << calc.Calculate(Triangle{6, 4})  << "\n";  // 12
